File: voice/sentence_buffer.py
"""Sentence buffer — accumulate streaming tokens and split on punctuation."""

from __future__ import annotations

# Chinese sentence enders (unambiguous, always split)
_HARD_BREAKS = frozenset("。！？；\n")

# English sentence enders (split only when followed by space or at buffer end)
_EN_BREAKS = frozenset(".!?")

# Soft breaks — only used when buffer gets long
_SOFT_BREAKS = frozenset("，、,")

_SOFT_THRESHOLD = 40   # chars before soft breaks activate
_FORCE_THRESHOLD = 80  # chars before forced split


class SentenceBuffer:
    """Accumulate streaming tokens and emit complete sentences.

    Usage::

        buf = SentenceBuffer()
        for token in stream:
            for sentence in buf.add(token):
                process(sentence)
        tail = buf.flush()
        if tail:
            process(tail)
    """

    def __init__(self) -> None:
        self._buf = ""

    def add(self, token: str) -> list[str]:
        """Add a token, return list of complete sentences (may be empty)."""
        self._buf += token
        sentences: list[str] = []

        while True:
            pos = self._find_break()
            if pos < 0:
                break
            sentence = self._buf[: pos + 1].strip()
            self._buf = self._buf[pos + 1 :]
            if sentence:
                sentences.append(sentence)

        # Force split if buffer exceeds threshold
        if len(self._buf) >= _FORCE_THRESHOLD:
            # Try to find a soft break for a cleaner cut
            cut = -1
            for i in range(len(self._buf) - 1, -1, -1):
                if self._buf[i] in _SOFT_BREAKS:
                    cut = i
                    break
            if cut >= 0:
                sentence = self._buf[: cut + 1].strip()
                self._buf = self._buf[cut + 1 :]
            else:
                sentence = self._buf.strip()
                self._buf = ""
            if sentence:
                sentences.append(sentence)

        return sentences

    def flush(self) -> str | None:
        """Return remaining buffered text, if any."""
        text = self._buf.strip()
        self._buf = ""
        return text or None

    def _find_break(self) -> int:
        """Find the earliest sentence break position, or -1."""
        for i, ch in enumerate(self._buf):
            # Chinese hard breaks — always split
            if ch in _HARD_BREAKS:
                return i
            # English breaks — split if followed by space or at end of buffer
            if ch in _EN_BREAKS and i >= 1:
                if i + 1 >= len(self._buf) or self._buf[i + 1] == " ":
                    return i

        # Soft breaks only when buffer is long enough
        if len(self._buf) > _SOFT_THRESHOLD:
            for i, ch in enumerate(self._buf):
                if ch in _SOFT_BREAKS:
                    return i

        return -1

File: public/modules/voice/sentence-buffer.js
/**
 * 分句器 — 从 voice/sentence_buffer.py 移植
 * 将流式文本切成适合 TTS 的句子
 */

const HARD_BREAKS = new Set(["。", "！", "？", "；", "\n"]);
const EN_BREAKS = new Set([".", "!", "?"]);
const SOFT_BREAKS = new Set(["，", "、", ","]);
const SOFT_THRESHOLD = 40;
const FORCE_THRESHOLD = 80;

export class SentenceBuffer {
  constructor() {
    this._buf = "";
  }

  /**
   * 添加文本 token，返回完成的句子数组
   * @param {string} token
   * @returns {string[]}
   */
  add(token) {
    this._buf += token;
    const sentences = [];

    // eslint-disable-next-line no-constant-condition
    while (true) {
      const idx = this._findBreak();
      if (idx < 0) break;
      const sentence = this._buf.slice(0, idx + 1).trim();
      this._buf = this._buf.slice(idx + 1);
      if (sentence) sentences.push(sentence);
    }

    // 强制截断过长的缓冲
    if (this._buf.length >= FORCE_THRESHOLD) {
      const sentence = this._buf.trim();
      this._buf = "";
      if (sentence) sentences.push(sentence);
    }

    return sentences;
  }

  /**
   * 冲刷缓冲区中的剩余文本
   * @returns {string|null}
   */
  flush() {
    const text = this._buf.trim();
    this._buf = "";
    return text || null;
  }

  _findBreak() {
    for (let i = 0; i < this._buf.length; i++) {
      const ch = this._buf[i];
      if (HARD_BREAKS.has(ch)) return i;
      if (EN_BREAKS.has(ch)) {
        // 英文句号后面必须是空格或结尾才算断句（避免缩写、小数）
        const next = this._buf[i + 1];
        if (!next || next === " " || next === "\n") return i;
      }
      if (SOFT_BREAKS.has(ch) && i >= SOFT_THRESHOLD) return i;
    }
    return -1;
  }
}
